core/textile/backup.go

```diff
@@ -0,0 +1,266 @@
+
+	if _, _, err = tc.UploadFileToHub(ctx, bucket, path, reader); err != nil {
+		log.Error(fmt.Sprintf("error backuping up file with reader (path=%+v b.Slug=%+v)", path, bucket.Slug()), err)
+		return err
+	}
+
+	if _, err = tc.setMirrorFileBackup(ctx, path, bucket.Slug()); err != nil {
+		log.Error(fmt.Sprintf("error setting mirror file as backup (path=%+v b.Slug=%+v)", path, bucket.Slug()), err)
+		return err
+	}
+
+	return nil
+}
+
+// backup a single file with a path provided
+func (tc *textileClient) backupFile(ctx context.Context, bucket Bucket, path string) error {
+
+	errc := make(chan error, 1)
+	pipeReader, pipeWriter := io.Pipe()
+
+	// go routine for piping
+	go func() {
+		defer close(errc)
+		defer pipeWriter.Close()
+
+		err := bucket.GetFile(ctx, path, pipeWriter)
+		if err != nil {
+			errc <- err
+			log.Error(fmt.Sprintf("error getting file (path=%+v b.Slug=%+v)", path, bucket.Slug()), err)
+			return
+		}
+	}()
+	if err := <-errc; err != nil {
+		return err
+	}
+
+	if err := tc.BackupFileWithReader(ctx, bucket, path, pipeReader); err != nil {
+		log.Error(fmt.Sprintf("error backing up file (path=%+v b.Slug=%+v)", path, bucket.Slug()), err)
+		return err
+	}
+
+	return nil
+}
+
+// backup all files in a bucket
+func (tc *textileClient) backupBucketFiles(ctx context.Context, bucket Bucket, path string) (int, error) {
+	var wg sync.WaitGroup
+	var count int
+
+	// XXX: we ignore errc (no atomicity at all) for now but we should return
+	// XXX: the errors, perhaps in a separate async call
+	errc := make(chan error)
+
+	dir, err := bucket.ListDirectory(ctx, path)
+	if err != nil {
+		return 0, err
+	}
+
+	wg.Add(len(dir.Item.Items))
+
+	for _, item := range dir.Item.Items {
+		if item.IsDir {
+			p := strings.Join([]string{path, item.Name}, "/")
+
+			n, err := tc.backupBucketFiles(ctx, bucket, p)
+			if err != nil {
+				return 0, err
+			}
+
+			count += n
+			continue
+		}
+
+		// parallelize the backups
+		go func(path string) {
+			defer wg.Done()
+
+			if err = tc.backupFile(ctx, bucket, path); err != nil {
+				errc <- err
+				return
+			}
+
+			count += 1
+		}(item.Path)
+	}
+
+	wg.Wait()
+
+	return count, nil
+}
+
+// meta, key and share with me thread ids
+func (tc *textileClient) userThreadIds(ctx context.Context, bucket Bucket) ([]thread.ID, error) {
+	dbIds := make([]thread.ID, 0)
+
+	// key
+	bucketThreadID, err := bucket.GetThreadID(ctx)
+	if err != nil {
+		return nil, err
+	}
+	dbIds = append(dbIds, *bucketThreadID)
+
+	// shared with me
+	publicShareId, err := tc.getPublicShareThread(ctx)
+	if err != nil {
+		return nil, err
+	}
+	dbIds = append(dbIds, publicShareId)
+
+	return dbIds, nil
+}
+
+// replicate meta, key and shared with me threads
+func (tc *textileClient) replicateThreadsToHub(ctx context.Context, bucket Bucket) (count int, err error) {
+	replicatedDbIds := make([]thread.ID, 0)
+
+	// poor man's transactionality
```

Comment:
I know threads have transactions capabilities backed in, maybe you can look into that so it would be easier to commit changes or rollback if there is an error.

Reply:
@perfectmak 

Great catch!

Unfortunately, we're calling https://github.com/textileio/go-threads/blob/8355a2e43aa1abc8b919851b3714abc0ce1d41c9/net/api/client/client.go#L200 under the hood so https://github.com/textileio/go-threads does not seem to offer transactions here, unless I'm misreading the code.